Applab: Google charts integration

islemaster opened the pull request:
Introduces a new "Chart" design mode element into Applab, and a new function `drawChartFromRecords()` that calls the Google charts API to render a chart with the requested parameters into a "Chart" element, using data from the Applab data API.

![screenshot from 2015-09-24 17 36 02](https://cloud.githubusercontent.com/assets/1615761/10090297/c4109b04-62e2-11e5-804a-abbe5a9d4ceb.png)
### Related links
- For those with access [here is the full spec for this feature.](https://docs.google.com/document/d/1qcqXcKTn2l6ZTgIotA6lKZfQXohHpmQWAiyTSBC4RNY/edit#)
- [Pie Chart API docs](https://developers.google.com/chart/interactive/docs/gallery/piechart)
- [Line Chart API docs](https://developers.google.com/chart/interactive/docs/gallery/linechart)
- [Bar Chart API docs](https://developers.google.com/chart/interactive/docs/gallery/barchart)
- [Scatter Chart API docs](https://developers.google.com/chart/interactive/docs/gallery/scatterchart)
# Features

We are introducing the feature with support for four chart types: Line, Bar, Pie and Scatter.

![screenshot from 2015-09-28 17 55 17](https://cloud.githubusercontent.com/assets/1615761/10152747/50318c9a-660a-11e5-961c-f81dc7613565.png)

![screenshot from 2015-09-28 17 55 39](https://cloud.githubusercontent.com/assets/1615761/10152750/526fdb4c-660a-11e5-90c4-f634586245c5.png)

Though not native to Google charts, we provide a loading spinner in the chart area while the asynchronous operation is in progress.  You can also see some of the charts API native interactivity in the animation below.

![chart-load-demo](https://cloud.githubusercontent.com/assets/1615761/10152763/747cb232-660a-11e5-9ce3-aaa0d277b0dc.gif)

Charts are added using the new Chart design mode element:

![drag-chart](https://cloud.githubusercontent.com/assets/1615761/10152806/e7b93720-660a-11e5-90ac-5676521acaa0.gif)

Then you call the following function to render the chart using data from one of your tables:

```
drawChart(chartId, chartType, tableName, columns[, options[, onComplete]])
```
- **chartId** is the element ID of the chart element
- **chartType** is one of ["bar", "line", "pie", "scatter"]
- **tableName** is a table name in the AppStorage API
- **columns** is an array of column names from the AppStorage API
- **options** is an object that gets passed through to the charts API directly.  Available options are [documented at Google](https://developers.google.com/chart/interactive/docs/gallery) (see also API docs linked near the top of this description).
- **onComplete** is a callback for when the chart is successfully rendered.
# Tests

There's a small suite of new unit tests for the `ChartApi` and its underlying `GoogleChart` types, especially around the sorts of errors and warnings it should generate with different inputs/datasets.

```
  ChartApi
    ✓ supports type BAR
    ✓ supports type PIE
    ✓ supports type LINE
    ✓ supports type SCATTER
    ✓ quotes and alphabetizes types for dropdown
    ChartType enum
      ✓ only contains supported types
      ✓ contains all supported types
    drawChartFromRecords
      ✓ returns a Promise
      ✓ rejects if element is not found
      ✓ rejects if element is wrong type
      ✓ rejects if chart type is not supported
      ✓ rejects if no columns array provided and no columns found in data
      ✓ rejects if no columns array provided and one column found in data
      ✓ warns and infers columns if no columns array provided and two columns found in data
      ✓ when inferring columns, prints list of all possible columns
      ✓ rejects if zero columns provided and less than two columns in data
      ✓ infers/warns if zero columns provided and two or more columns in data
      ✓ rejects if only one column provided and less than two columns in data
      ✓ infers/warns if only one column provided and two or more columns in data
      ✓ warns about empty dataset
      ✓ does not warn about empty dataset when given data
      ✓ warns about empty column
      ✓ does not warn about empty column if all columns have data
      ✓ pie charts warn about three columns
      ✓ bar charts do not warn about three columns
      ✓ line charts do not warn about three columns
      ✓ scatter charts do not warn about three columns
    inferColumnsFromRawData
      ✓ extracts all columns from data
      ✓ adds columns in the order they are encountered
```

I also added a test for the design element (basic dragging to work area, resizing, deleting).
# Other improvements
- I'm pulling in an [ES6 Promise polyfill](https://github.com/jakearchibald/es6-promise) because I had a use case for `Promise.all`.  Hopefully we'll find this useful in the future too (I know I will).
- `Applab.getIdDropdown` now accepts a few types of selectors: In addition to tag names, you can filter by `.class-name` or `#identifier`.  I introduced some tests for this method since it's more complex than before.
# TODO
- [x] Test various warnings for `drawChartFromRecords()`
- [x] Only load Google jsapi in ~~Applab~~ droplet levels
-  ~~Clean up experience of adding function from design mode.~~ TBD later.
- [x] Loading spinner in chart div between `drawChart...` call and rendering.
- [x] Custom min and max columns per chart type.
-  ~~Maybe break into onSuccess and onError callbacks~~ TBD later.
### For later:
- Provide a `drawChart` function that does not hit the data API (takes array of data directly).
- Full documentation for new function (volunteer-driven).

islemaster commented on dashboard/app/views/layouts/application.html.haml:
```diff
@@ -19,6 +19,8 @@
     = render partial: 'layouts/analytics'
     = render partial: 'layouts/javascript_strings'
+    - # TODO: Only load jsapi when needed (in Applab)
+    %script{type: 'text/javascript', src: 'https://www.google.com/jsapi'}
```
Opinions on doing this include cleanly?

islemaster commented:
Partial improvement: We're now loading the Google loader only in droplet levels.  We'll probably make this more precise to loading as needed during the optimization task this sprint, so I think it's sufficient to get loading out of the haml template.

islemaster commented:
@davidsbailey I'm ready for review!  Please take a look tomorrow, and let me know if you have any questions.

sfilman commented:
This reminded me of something interesting...when we first introduce design
mode in the curriculum should we be hiding the chart block in design mode?
After all, the drawChartFromRecords block won't even be exposed. I don't
think it's necessary to handle right now, but something we might want to do
later. It may even be interesting for perf one day-- don't load the google
stuff unless the drawChartFromRecords is in the toolbox.

On Mon, Sep 28, 2015 at 6:07 PM, Brad Buchanan notifications@github.com
wrote:

> @davidsbailey https://github.com/davidsbailey I'm ready for review!
> Please take a look tomorrow, and let me know if you have any questions.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/code-dot-org/code-dot-org/pull/4184#issuecomment-143915540
> .

davidsbailey commented:
@sfilman , it sounds like the feature you'd like is for the design mode toolbox to be customizable per-level, is that correct?

sfilman commented:
I wasn't thinking that exactly, though that is an interesting idea. I was
thinking that if the google charts stuff in particular is "heavy" then we
could be smart enough to know whether we've even exposed the data blocks in
the toolbox, and if not, the google chart block and the google stuff that
comes with it don't necessarily need to be there.

On Tue, Sep 29, 2015 at 9:01 AM, davidsbailey notifications@github.com
wrote:

> @sfilman https://github.com/sfilman , it sounds like the feature you'd
> like is for the design mode toolbox to be customizable per-level, is that
> correct?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/code-dot-org/code-dot-org/pull/4184#issuecomment-144103714
> .

davidsbailey commented:
L:eyes:king

davidsbailey commented:
@sfilman if the goal is to avoid loading the code for charts only, then it's possible we could accomplish that with a per-level flag, e.g. `hideCharts`. However I am not sure a per-level flag will help us in the HoC share scenario, unless we decided to disable Charts entirely for whatever App Lab free-play level we are promoting for HoC. 

By contrast, hiding all of design mode is easier for share links at app init time, since the design mode UI code is never needed in share mode.

If we did implement just-in-time loading of javascript code (rather than making all decisions at app init time based on level config), then we could be smarter about not loading charts unless/until it is needed.

Thanks for bringing this up, we'll have to keep Charts in mind for potential performance improvements.

islemaster commented:
@davidsbailey and @sfilman I also want to point out that the only thing we always load in Applab is a small Google loader, not the whole charts library.  We actually load requirements for the specific requested chart type as late as possible, at runtime - and my experience so far suggests that even those almost load faster than our own hosted dependencies.  We can keep this in mind, but there are likely to be bigger issues to tackle first.

davidsbailey commented on apps/src/applab/commands.js:
```diff
@@ -1388,3 +1389,132 @@ applabCommands.getUserId = function (opts) {
+  apiValidateType(opts, 'drawChartFromRecords', 'chartId', opts.chartId, 'string');
+  apiValidateType(opts, 'drawChartFromRecords', 'chartType', opts.chartType, 'string');
+  apiValidateType(opts, 'drawChartFromRecords', 'tableName', opts.tableName, 'string');
+  apiValidateType(opts, 'drawChartFromRecords', 'columns', opts.columns, 'array');
```
columns should be optional, right?

davidsbailey replied:
I see that the spec says `columns` is required, but I also see that you added logic to infer columns when they are not specified, so I'm not quite sure what the desired behavior is here.

islemaster replied:
This is me anticipating the next step (maybe too much) which is that we'll want a `drawChart(chartId, chartType, data[, options[, callback]])` method.  For that one I think we'll want to infer columns.

davidsbailey commented on apps/src/applab/GoogleChart.js:
```diff
@@ -0,0 +1,463 @@
+GoogleChart.prototype.loadData = function (rawData, columnList) {
+  // If columnList is not provided, infer it from raw data
+  if (columnList === undefined) {
+    columnList = GoogleChart.inferColumnsFromRawData(rawData);
```
you probably want to omit the `id` field from this list

islemaster replied:
Oooh... Yes I do, _if_ we are loading AppStorage API table data, which this layer knows nothing about.  For now, while `columns` is required for `drawChartFromRecords` I'm not sure that's helpful, but if we ever make it optional we may want to scrub "id" from the data up at the `ChartApi` level.

davidsbailey commented on apps/src/applab/GoogleChart.js:
```diff
@@ -0,0 +1,463 @@
+ */
+GoogleChart.prototype.loadData = function (rawData, columnList) {
+  // If columnList is not provided, infer it from raw data
+  if (columnList === undefined) {
```
What is the desired behavior when columnList is empty? It seems like we might want to print a warning rather than displaying an empty chart (assuming that's what happens when you pass Google Charts an empty list of columns).

islemaster replied:
We actually raise an error if there are less than two columns, [implemented here](https://github.com/code-dot-org/code-dot-org/pull/4184/files#diff-cfb0ef5d8d2609de79b1510822620cc3R179).

islemaster commented on apps/src/applab/ChartApi.js:
```diff
@@ -0,0 +1,217 @@
+      chart.loadDependencies(),
+      this.fetchTableData_(tableName)
+    ]).then(function (results) {
+      return chart.loadData(results[1], columns);
```
...right here.  If `columns` is not defined, I should strip "id" from the table data here?

davidsbailey replied:
That's better but still not great, because then students are prevented from explicitly including `id` as a column. I would recommend extracting the call to inferColumnsFromData out of loadData and into `ChartApi.prototype.drawChartFromRecords` here.

islemaster commented:
Discussion with Dave:

Students will have a hard time figuring out what column names they should use.  If they omit columns or put bad columns in, we should infer a couple of good columns and print a warning with valid column options from the returned dataset.

Eventually table and column dropdowns in block mode would be awesome.

davidsbailey commented on apps/src/applab/GoogleChart.js:
```diff
@@ -0,0 +1,463 @@
+  PieChart.superPrototype.verifyData_.call(this, data, columns);
+
+  if (columns.length > 2) {
+    this.warn('Too many columns for pie chart; only using the first 2.');
```
We need to give this same warning for scatter charts too.

islemaster replied:
Why? You can use _N_ columns to represent _N-1_ series on a scatter chart ([API reference](https://developers.google.com/chart/interactive/docs/gallery/scatterchart?hl=en#data-format)).

davidsbailey replied:
You are correct! I based this off of my own comment in the spec which I see now is wrong.

davidsbailey commented on apps/test/ChartApiTest.js:
```diff
@@ -0,0 +1,367 @@
+          }));
+    });
+
+    it ("rejects if no columns array provided", function (testDone) {
```
I would propose changing this line to "rejects if no columns array provided and no columns found in the data". Otherwise, this test is confusing because it suggests that `columns` must always be present, whereas in reality `columns` do not need to be present if they can be inferred from the data.

islemaster replied:
:+1:

davidsbailey commented on apps/test/ChartApiTest.js:
```diff
@@ -0,0 +1,367 @@
+          }));
+    });
+
+    it ("does not warn about empty dataset when given data", function (testDone) {
```
How about adding a similar test to this one but without the `columns` parameter, which does `assertNotWarns(..., "Not enough columns...")` ? To verify that columns can be inferred.

islemaster replied:
:+1: although I may want to change this assertion based on what we talked about (warning and displaying possible column names).

davidsbailey commented on apps/src/applab/applab.js:
```diff
@@ -1616,30 +1616,75 @@ Applab.getAssetDropdown = function (typeFilter) {
-    };
-  });
+/**
+ * Internal version of getIdDropdownFromDom_, which takes a documentRoot
```
I don't understand the beginning of this comment, since there is no corresponding external version of getIdDropdownFromDom_. "Internal version of getIdDropdownFromDom_" --> "Internal helper for getIdDropdown", perhaps?

islemaster replied:
Comment bugs `//`:bug::beetle::bug: nice catch

davidsbailey commented on apps/src/applab/applab.js:
```diff
@@ -1616,30 +1616,75 @@ Applab.getAssetDropdown = function (typeFilter) {
+      divApplabChildren.children().toArray());
+
+  return elements
+      .filter(makeElementFilterFunction(filterSelector))
```
I am scared of rolling our own element selector in production when jQuery can do it for us.  jQuery works in our test environment, right? so can we just do: `$(elements).filter(filterSelector)` ?

islemaster replied:
Sounds good.  Have we officially decided we're cool with jQuery in Applab?  I thought we were avoiding it, but now I see it already exists in this file.

davidsbailey replied:
A lot of applab was written without jQuery, partially because our `grunt dev` and mochaTest environments couldn't handle it. Since those have been fixed, jQuery in applab is allowed and encouraged. @Bjvanminnen can chime in if he disagrees.

Bjvanminnen replied:
jQuery should be fine.

davidsbailey commented on shared/js/loadApp.js:
```diff
@@ -50,7 +68,8 @@ module.exports = function (callback) {
         .then(loadSource('droplet/droplet-full'))
-        .then(loadSource('tooltipster/jquery.tooltipster'));
+        .then(loadSource('tooltipster/jquery.tooltipster'))
+        .then(loadExternalSource('https://www.google.com/jsapi'));
```
do we have any idea how much code this is loading? it looks like it makes a bunch of other calls to obfuscated members of `google.loader`.

islemaster replied:
It looks like this loads 5.8 KB of JavaScript, which stacks up favorably against dependencies like ace (330 KB minified) or just the qTip jQuery extension (43.5 KB minified).  It's close to md5.js at 6.5 KB.

I can't find evidence that it does anything else until you invoke `google.load` to pull in packages you need, which we do right before rendering the chart.

![screenshot from 2015-09-30 11 22 21](https://cloud.githubusercontent.com/assets/1615761/10203656/2dbed27c-676c-11e5-9211-c2e3703ac690.png)

davidsbailey commented on apps/test/applabTest.js:
```diff
@@ -48,3 +49,71 @@ describe('designMode.addScreenIfNecessary', function () {
+    ]);
+  });
+
+  it('does not accidentally pick up superset classes', function () {
```
FWIW, this test case is an example of something we wouldn't have to worry about if we rely on jQuery to filter elements for us.

islemaster replied:
FYI, I'm leaving these tests in even though we're switching to the jQuery implementation.  It never hurts to prove that your libraries do what you expect them to :wink:.

davidsbailey commented:
Great work Brad! This is a huge, awesome feature and the approach you've taken looks great. My only real concern is the reimplementation of jQuery element selectors mentioned [here](https://github.com/code-dot-org/code-dot-org/pull/4184#discussion_r40752412). If you want to get this checked in on Weds, then let's reach a resolution on that issue before I disappear for classroom visit at noon.

islemaster commented:
I've moved the task of inferring columns up into `ChartApi` and simplified the `GoogleChart` interface again (it no longer requires a separate `loadData` call).  Improved the warnings we provide when we have to infer columns (see tests) to include a list of valid column names. Also using jQuery for element selection now.

@davidsbailey Anything else?

davidsbailey commented:
LGTM
